lib/stellar/sep10.rb

```diff
@@ -0,0 +1,383 @@
+        else
+          client_signers.add(signer)
+        end
+      end
+
+      if client_signers.empty?
+        raise InvalidSep10ChallengeError.new("At least one signer with a G... address must be provied") 
+      end
+
+      # verify all signatures in one pass
+      all_signers = client_signers + Set[server.address]
+      signers_found = verify_tx_signatures(
+        tx_envelope: te, signers: all_signers
+      )
+
+      # ensure server signed transaction and remove it
+      if !signers_found.delete?(server.address)
+        raise InvalidSep10ChallengeError.new("Transaction not signed by server: #{server.address}")
+      end
+
+      # Confirm we matched signatures to the client signers.
+      if signers_found.empty?
+        raise InvalidSep10ChallengeError.new("Transaction not signed by any client signer.")
+      end
+
+      # Confirm all signatures were consumed by a signer.
+      if signers_found.length != te.signatures.length - 1
+        raise InvalidSep10ChallengeError.new("Transaction has unrecognized signatures.")
+      end
+
+      return signers_found
+
+    end
+
+    Contract(C::KeywordArgs[
+      challenge_xdr: String,
+      server: Stellar::KeyPair,
+      threshold: Integer,
+      signers: SetOf[::Hash],
+    ] => C::SetOf[::Hash])
+    # Verifies that for a SEP 10 challenge transaction all signatures on the transaction 
+    # are accounted for and that the signatures meet a threshold on an account. A 
+    # transaction is verified if it is signed by the server account, and all other 
+    # signatures match a signer that has been provided as an argument, and those 
+    # signatures meet a threshold on the account.
+    #
+    # @param challenge_xdr [String] SEP0010 transaction challenge transaction in base64.
+    # @param server [Stellar::KeyPair] keypair for server's account.
+    # @param threshold [Integer] The medThreshold on the client account.
+    # @param signers [SetOf[::Hash]]The signers of client account.
+    #
+    # @return [SetOf[::Hash]]
+    #
+    # Raises a InvalidSep10ChallengeError if:
+    #   - The transaction is invalid according to Stellar::SEP10.read_challenge_transaction.
+    #   - One or more signatures in the transaction are not identifiable as the server 
+    #     account or one of the signers provided in the arguments.
+    #   - The signatures are all valid but do not meet the threshold.
+    def self.verify_challenge_tx_threshold(
+      challenge_xdr:,
+      server:,
+      threshold:,
+      signers:
+    )
+      signer_str_set = signers.map { |s| s['key'] }.to_set
+      signer_strs_found = verify_challenge_tx_signers(
+        challenge_xdr: challenge_xdr, 
+        server: server, 
+        signers: signer_str_set
+      )
+  
+      weight = 0
+      signers_found = Set.new
+      signers.each do |s|
+        if !signer_strs_found.include?(s['key'])
+          next
+        end
+        signers_found.add(s)
+        weight += s['weight']
+      end
```

Comment:
I think these lines result in double counting a weight because it is iterating over the input `signers` that may be an array containing duplicates. We should have a test that catches this case, specifically duplicate signers in the input list shouldn't cause an error but should only be counted once. We have a test that captures it for the `_signers` function but not this function and the bug exists in this function.

Reply:
I added code to remove the signer key from `signer_strs_found` before adding its weight to the total. So if there is a duplicate signer in `signers`, it will be skipped the second time it appears in the loop.